README.md
# Harness for pysdl2

These are a set of classes to make easier to use [pysdl2](https://pysdl2.readthedocs.org).

This is a work in progress so use it at your own risk!


## Required

 - Python 3
 - pysdl2
 - SDL2 and SDL2\_Mixer installed in your system
 - Optionally SDL2\_Image (otherwise only uncompressed BMP images are supported)


## How does it looks like?

It is inspired by [pyglet](http://www.pyglet.org/) and specially focused
on 2D games.

Example:

```python
#!/usr/bin/env python

from harness import Harness

game = Harness(width=320, height=240, zoom=3)

title = game.load_resource("title.bmp")

@game.draw
def draw(renderer):
	renderer.draw(title, 10, 10)

@game.update
def update(dt):
	if game.keys[game.KEY_ESCAPE]:
		game.quit()

game.loop()
```

See `harness` module docstrings for a complete list of classes and methods.

There's an example game in `example.py` and remember that you can still use
pysdl2 directly if you need to!

## Components

Harness tries to provide a clean and simple interface to the following
components:

 1. The game loop
 2. Resource management
 3. Controls
 4. Audio

### 1. The game loop

Harness implements a game loop with a fixed frame rate determined by the vsync
of the screen (usually 60 FPS), with support for fixed updates for the game
logic (by default at 80 times per second).

The usual workflow is:

 1. Create a Harness object (we'll call it `game` in the examples).
 2. Load resources.
 3. Declare the draw and update functions.
 3. Run the `loop()` method in your Harness instance.

The game loop should be called once and it will run until the game is quitted
(eg, using `quit()` method).

Draw functions can be defined with the `draw` decorator, and update
functions with the `update` decorator.

The draw functions should expect a "renderer" parameter that allows to draw
textures, bitmap fonts, etc.

Example:

```python

game = Harness()

tex = game.load_resource("bitmap.bmp")

@game.draw
def draw(renderer):
    renderer.draw(tex)

game.loop()
```

The update function should expect a "dt" parameter that provides the delta
time (time elapsed between updates); in this case fixed at `Harness.UFPS_DT`
(1 / UFPS).

Example:

```python

game = Harness()

@game.update
def update(dt):
    print("%s elapsed since last update" % dt)

game.loop()
```

Several draw and update functions can be defined and they will be run in the
same order they were defined.

The game instance can be accessed from the update function to test for key
states, quit the game, etc.

The method `quit()` can be used to exit the game loop.

Example:
```python

game = Harness()

@game.update
def update(dt):

    if game.keys[game.KEY_ESCAPE]:
        game.quit()
        # in case we don't want to complete the update
        return

game.loop()
```

A draw or update function can be removed from the game loop with `remove_handler()`
method, passing the function to be removed as parameter.

Example:
```python

game = Harness()
debug = False

def update_debug(dt):
	print(dt)

@game.update
def update(dt):
    global debug

    if game.keys[game.KEY_D]:
        print("D was pressed!")
        if debug:
            # remove the update_debug update function
            game.remove_handler(update_debug)
        else:
            # add a new update function
            game.update(update_debug)
        debug = not debug
        # remove the key press once processed
        game.keys[game.KEY_D] = False

    if game.keys[game.KEY_ESCAPE]:
        game.quit()

game.loop()

```

### 2. Loading resources

Resources can loaded with `load_resource()` method. This method allows loading
resources searching for them in the paths specified in the `resource_path` list.

By default the files will be searched for in the "data" subdirectory at the same
level as the script running the game.

Depending on the resource some extra libraries may be required in the system
(eg, `SDL_Image`).

Resources not in use can be freed using `free_resources()` method, but
be careful to not use any reference to the resource once it has been released.

Harness will free all resources after exiting the game loop.

#### 2.1 Bitmap fonts

The method `load_bitmap_font()` can be used to load a image that will be used to draw
text with `renderer.draw_text()`. Harness will map a text string into a fixed
width and height part of the font image.

Example:
```python

game = Harness()

font = game.load_bitmap_font("font.png", width=6, height=10)

@game.draw
def draw(renderer):
    renderer.draw_text(font, 10, 10, "This is a text!")

game.loop()
```

Fonts can be freed with `free_resources()`.

### 3. Controls

The state of the keys is exposed in `keys` dictionary and it
gets updated in each game loop iteration.

In `Harness.KEY_*` there are constants to test in the "keys" dictionary. If a key
is being pressed, the value in the dictionary will be `True`.

Example:
```python

game = Harness()

@game.update
def update(dt):

    if game.keys[game.KEY_ESCAPE]:
        game.quit()

game.loop()
```

#### 3.1 Game controllers

Game controllers can be mapped into key states so the game can access to the
controller like the player was using the keyboard.

The default mapping is:

 - DPad up: up arrow key
 - DPad down: down arrow key
 - DPad left: left arrow key
 - DPad right: right arrow key
 - Button A: key c
 - Button B: key v
 - Start button: key s
 - Back button: escape key

Harness will manage the controller automatically in the game loop updating the
`keys` dictionary as needed. 

`has_controllers` property can be checked to see if any game controller was
detected. Harness includes a game controller database with definitions for most
common devices, and SDL2 functions can be used to add more. If there's no information
about a given controller, it will be silently ignored.

In order to use a controller, the `controllers` property can be accessed to
activate any detected controller.

Example:
```python

game = Harness()

# enumerate all detected controllers
for controller in game.controllers:
    print(controller.name)

```

Once the controller has been activated, it can be deactivated using `close()`
controller method.

The key mapping can be changed using the `set_mapping()` method on the controller.

Example:
```python

game = Harness()

# first controller
controller = game.controllers[0]

# remap button a to key a
controller.set_mapping(a="KEY_A")
```

The valid parameters are: up, down, left, right, a, b, start and back. Use a
string defining the key (see `Harness.KEY_*`).

The use of a controller won't disable the keyboard. If that is required, the
game controllers can be accessed using SDL2 functions directly.

### 4. Audio

The method `play()` can be used to play a sample loaded with `load_resource()`.
Optionally a `loops` parameter can be provided stating how many times the sample
will be repeated (use -1 for an infinite loop).

By default .ogg and .wav files are supported (in theory it could load any
format supported by `SDL_Mixer` but Harness will only identify files with the
aforementioned extensions).

`play()` returns the channel number used to play the sample and that
number can be used to muted the channel with `stop_playback()` (if a channel
number s not provided, it will stop all channels).

By default `Harness.AUDIO_CHANNELS` channels are allocated (6 channels).

## Using OOP

Harness can be used in a class to take advantage of object oriented programming
and avoid the use of global variables. Just use composition and register the
update and draw methods with `update()` and `draw()` instead of using the
decorators:

Example:
```python
from harness import Harness

class MyGame(object):

    def __init__(self):
        self.harness = Harness()

        # register update and draw methods
        self.harness.update(self.update)
        self.harness.draw(self.draw)

        # load some resources
        self.image = self.harness.load_resource("image.png")

    def run(self):
        self.harness.loop()

    def update(self, dt):
        if self.harness.keys[self.harness.KEY_ESCAPE]:
            self.harness.quit()

    def draw(self, renderer):
        renderer.draw(self.image)


if __name__ == "__main__":
    game = MyGame()
    game.run()
```

See `example-oop.py`.

## Author

Juan J. Martinez <jjm@usebox.net>

This is free software under MIT license terms.



CHANGES
0.2 (2015-05-31)
----------------

 - Python 2.7.x compatibility
 - README converted to reStructuredText (rst)

0.1.1 (2015-05-31)
------------------

 Only fixes long description for PyPi, no other changes.

0.1 (2015-05-31)
----------------

 First public release!



example.py
#!/usr/bin/env python
"""
Harness demo, a game example.

Tested with Python 3.4, may or may not work with Python 2!
"""

from random import randint, shuffle

from harness import Harness

game = Harness(title="pysdl2 HARNESS demo", width=240, height=240, zoom=3)

if game.has_controllers:
    # use first game controller; harness will manage this internally
    # but it's useful to keep a reference (eg, to set the keyboard mapping,
    # get the controller name, etc)
    controller = game.controllers[0]

background = game.load_resource("background.png")
title = game.load_resource("title.png")
font = game.load_bitmap_font("font.png", width=6, height=10)

# load all the sprites in just one image so they're in one big texture
tiles = game.load_resource("tiles.png")

game.set_icon("icon.png")

dance = game.load_resource("harness-dance.ogg")
dance_hurry = game.load_resource("harness-dance-hurry.ogg")
gameover = game.load_resource("gameover.ogg")
hurryup = game.load_resource("hurryup.ogg")
time = game.load_resource("time.ogg")

scenes = []
hiscore = 0

class MenuScene(object):

    # only one direction
    dragon_frames = ((0, 24, 24, 24), (24, 24, 24, 24))
    knight_frames = ((48, 48, 24, 24), (72, 48, 24, 24))

    def __init__(self):
        self.intro_channel = None
        self.counter = 2
        self.title_y = -100

        self.anim_delay = 0
        self.frame = 0

        self.dragon = [tiles.get_texture(*frames) for frames in self.dragon_frames]
        self.knight = [tiles.get_texture(*frames) for frames in self.knight_frames]

        self.boing = game.load_resource("boing.ogg")
        game.play(self.boing)

    def draw(self, renderer):
        renderer.draw(background)
        renderer.draw(title, dest_rect=(0, int(self.title_y), 240, 60))

        # wait until the title is in place
        if self.title_y >= 40:
            if 2 < self.counter < 12:
                renderer.draw_text(font, 120, 132, "Press 's' to start!", align="center")

            renderer.draw_text(font, 120, 10, "Copyright (c) 2015 usebox.net", align="center")
            renderer.draw_text(font, 120, 48, "HI: %04i" % hiscore, align="center")

            renderer.draw_text(font, 120, 196, "Use the arrows to move", align="center")
            renderer.draw_text(font, 120, 208, "and collect the goodies!", align="center")

            renderer.draw_text(font, 120, 106, "a game by @reidrac",
                    align="center", tint=(98, 100, 220, 255))

            # draw the animation cycle; frame 1 will be drawn 1 pixel higher
            renderer.draw(self.dragon[self.frame],
                          x = 12,
                          y = 192 - self.frame,
                          )
            renderer.draw(self.knight[self.frame],
                          x = 204,
                          y = 192 - self.frame,
                          )

    def update(self, dt):

        # initial animation (title falls into place)
        if self.title_y < 40:
            self.title_y += dt * 120
            return
        elif self.boing:
            # won't be used again
            game.free_resource("boing.ogg")
            self.boing = None
        elif self.intro_channel is None:
            # loop the intro music
            self.intro_channel = game.play(dance, loops=-1)

        # the "start" will blink
        self.counter += dt * 10
        if self.counter > 12:
            self.counter -= 12

        # frame animation
        self.anim_delay += dt * 10
        if self.anim_delay > 1.5:
            self.anim_delay = 0
            self.frame = 0 if self.frame else 1

        # controls
        if game.keys[game.KEY_ESCAPE]:
            game.quit()
        elif game.keys[game.KEY_S]:
            # press "s" to play
            game.stop_playback(self.intro_channel)
            self.intro_channel = None
            scenes.append(ReadyScene())
            return

class ReadyScene(object):

    def __init__(self):
        self.delay = 16

    def draw(self, renderer):
        renderer.draw(background)
        renderer.draw_text(font, 120, 100, "READY?", align="center")

    def update(self, dt):
        if self.delay > 0:
            self.delay -= dt * 10
            if self.delay <= 0:
                scenes.pop()
                scenes.append(PlayScene())

class GameOverScene(object):

    def __init__(self):
        self.delay = 80

        # play it once
        game.play(gameover)

    def draw(self, renderer):
        renderer.draw(background)
        renderer.draw_text(font, 120, 100, "GAME OVER", align="center")

    def update(self, dt):
        if self.delay > 0:
            self.delay -= dt * 10
            if self.delay <= 0:

                # back to menu
                scenes.pop()

class PlayScene(object):

    BW = 7
    BH = 7

    MAX_TILES = 10
    TILES = 8

    def __init__(self):
        self.stage = 1
        self.score = 0

        # subtextures for the board tiles
        self.tiles = [tiles.get_texture(*tuple([i * 24, 0, 24, 24])) for i in range(self.MAX_TILES)]

        self.music_channel = game.play(dance, loops=-1)
        self.next_stage()

    def next_stage(self):
        self.time = 12
        self.hurry_up = None
        self.time_tint = None
        self.game_over = None
        self.prev_time = 0

        # generate a random board
        tileset = [range(self.MAX_TILES)]
        shuffle(tileset)
        self.board = [randint(0, self.TILES - 1) for i in range(self.BW * self.BH)]

    def draw(self, renderer):
        renderer.draw(background)

        renderer.draw_text(font, 4, 4, "SCORE %04i" % self.score)
        renderer.draw_text(font, 236, 4, "STAGE %i" % self.stage, align="right")

        if self.hurry_up:
            # blink a warning when the time is running out
            if int(self.hurry_up) & 1:
                renderer.draw_text(font, 120, 9, "HURRY UP!", align="center")
        else:
            renderer.draw_text(font, 120, 9, "TIME: %02i" % int(self.time), align="center", tint=self.time_tint)

        # draw the board
        for y in range(self.BH):
            for x in range(self.BW):
                renderer.draw(self.tiles[self.board[x + y * self.BW]],
                              x=(120 - (24 * self.BW // 2)) + x * 24,
                              y=(120 - (24 * self.BH // 2)) + y * 24,
                              )

    def update(self, dt):

        # HURRY UP! delay
        if self.hurry_up:
            self.hurry_up -= dt * 10

            if self.hurry_up <= 0:
                self.hurry_up = 0
                if self.music_channel is None:
                    self.music_channel = game.play(dance_hurry, loops=-1)
            else:
                return

        self.time -= dt

        # set HURRY UP once
        if int(self.time) == 10 and self.hurry_up is None:
            self.hurry_up = 12
            self.time_tint = (255, 0, 0, 255)
            self.prev_time = 10
            game.play(hurryup)
            game.stop_playback(self.music_channel)
            self.music_channel = None
            return

        # beep on the last seconds
        if self.prev_time and int(self.time) != self.prev_time:
            self.prev_time = int(self.time)
            game.play(time)

        # set GAME OVER
        if int(self.time) == 0:

            # stop music
            if self.music_channel is not None:
                game.stop_playback(self.music_channel)

            global hiscore
            if self.score > hiscore:
                hiscore = self.score

            scenes.pop()
            scenes.append(GameOverScene())
            return

        # controls
        if game.keys[game.KEY_ESCAPE]:
            # avoid leaving the game just after
            # leaving this scene!
            game.keys[game.KEY_ESCAPE] = False

            # stop music
            if self.music_channel is not None:
                game.stop_playback(self.music_channel)

            # go to previous scene
            scenes.pop()
            return

@game.draw
def draw(renderer):
    # draw last scene
    scenes[-1].draw(renderer)

@game.update
def update(dt):
    # uddate last scene
    scenes[-1].update(dt)

scenes.append(MenuScene())

game.loop()



example-oop.py
#!/usr/bin/env python
from harness import Harness

class MyGame(object):

    def __init__(self):
        self.harness = Harness(width=240, height=240, zoom=3)

        # register update and draw methods
        self.harness.update(self.update)
        self.harness.draw(self.draw)

        # load some resources
        self.font = self.harness.load_bitmap_font("font.png", width=6, height=10)

        self.debug = False

    def run(self):
        self.harness.loop()

    def update_debug(self, dt):
        print(dt)

    def update(self, dt):
        if self.harness.keys[self.harness.KEY_D]:
            print("***")
            self.harness.keys[self.harness.KEY_D] = False
            self.debug = not self.debug
            if self.debug:
                self.harness.update(self.update_debug)
            else:
                self.harness.remove_handler(self.update_debug)

        if self.harness.keys[self.harness.KEY_ESCAPE]:
            self.harness.quit()

    def draw(self, renderer):
        renderer.draw_text(self.font, 120, 120, "Press d to toggle debug update,", align="center")
        renderer.draw_text(self.font, 120, 130, "ESC to quit", align="center")


if __name__ == "__main__":
    game = MyGame()
    game.run()



harness.py
#!/usr/bin/env python
"""
harness.Game,
some simple classes to make working with pysdl2 easier.

Copyright (C) 2015 by Juan J. Martinez <jjm@usebox.net>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
"""
import sys
import os
import ctypes
from collections import namedtuple

version = "0.1-alpha"

try:
    import sdl2
except ImportError as ex:
    sys.exit("SDL2 library not found: %s" % ex)

try:
    from sdl2 import sdlmixer
except ImportError as ex:
    sys.exit("SDL2_Mixer library not found: %s" % ex)


class Game(object):
    """
    Game object.

    Parameters:

        title: windows title
        width: with in pixels of the draw area
        height: height in pixels of the draw area
        zoom: scale up the output, or use 1 to disable

    The Game loop

        Call it once and it will run until the game is quitted (eg, using
        "quit()" method).

        It implements fixed draw (cap a FPS) and fixed updates.

        Define draw functions with the "game.draw" decorator, and update
        functions with the "game.update" decorator.

        The draw functions should expect a "renderer" parameter that
        allows to draw textures and bitmap fonts.

        Example:

        ```python

            game = Game()

            tex = game.load_resource("bitmap.bmp")

            @game.draw
            def draw(renderer):
                renderer.draw(tex)

            game.loop()
        ```

        The update functions should expect a "dt" parameter that provides
        the delta time (time elapsed between updates); in this case fixed
        at UPDATE_DT (1 / UPDATE_FPS).

        Example:

        ```python

            game = Game()

            @game.update
            def update(dt):
                print("%s elapsed" % dt)

            game.loop()
        ```

        Several draw and update functions can be defined and they will be
        run in the same order they were defined.

        The game instance can be accessed from the update function to
        test for key states, quite the game, etc.

    Loading resources

        This method allows loading resources searching for them
        in the paths specified in game.resource_path.

        By default this files will be searched in the "data"
        subdirectory at the same level as the script running
        the game.

        Depending on the resource some extra libraries may be
        required in the system (eg, SDL_Image).

        Resources not in use can be freed using game.free_resources(filename)
        method.

    Bitmap fonts

        game.load_bitmap_font can be used to load a image that will be
        used to draw text with renderer.draw_text.

        Example:
        ```python

            game = Game()

            font = game.load_bitmap_font("font.png", width=6, height=10)

            @game.draw
            def draw(renderer):
                renderer.draw_text(font, 10, 10, "This is a text!")

            game.loop()
        ```

        Fonts can be freed with game.free_resources.

    Audio

        game.play can be used to play a sample loaded with game.load_resource.
        Optionally a "loops" parameter can be provided stating how many times
        the sample will be repeated (use -1 for an infinite loop).

        game.play returns the channel number used to play the sample and that
        channel can be muted with game.stop_playback (don't provide a channel
        number to stop all channels).

        By default Game.AUDIO_CHANNElS channels are allocated.

    Useful properties

        Keyboard

        keys: updated in each update loop with the status of all the keys.
        KEY_*: constants to test in the "keys" dictionay.

        Resources

        resource_path: list of paths to search for resources.

    Useful methods

        quit(): quit the game.

        Example:
        ```python

            game = Game()

            @game.update
            def update(dt):

                if game.keys[game.KEY_ESCAPE]:
                    game.quit()

            game.loop()
        ```
    """
    # the update functions will be called a fixed number of FPS
    UPDATE_FPS = 80
    UPDATE_DT = 1.0 / UPDATE_FPS

    # draw frames per second
    FPS = 60
    DRAW_DT = 1.0 / FPS

    FONT_MAP = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!?()@:/'., "

    AUDIO_CHANNELS = 6

    def __init__(self, title=None, width=320, height=200, zoom=1):

        self.title = title.encode() or b"SDL2 Game"
        self.width = width
        self.height = height
        self.zoom = zoom

        self._quit = False
        self.dt = 0
        self.update_handlers = []
        self.draw_handlers = []
        self.resource_path = [os.path.join(os.path.dirname(os.path.realpath(__file__)), "data"),]
        self.resources = {}
        self.bitmap_fonts = {}

        for attr in dir(sdl2):
            if attr.startswith("SDL_SCANCODE_"):
                setattr(self, attr.replace("SDL_SCANCODE_", "KEY_"), getattr(sdl2, attr))

        sdl2.SDL_Init(sdl2.SDL_INIT_VIDEO|sdl2.SDL_INIT_AUDIO)
        sdlmixer.Mix_Init(sdlmixer.MIX_INIT_OGG)
        sdlmixer.Mix_OpenAudio(44100, sdlmixer.MIX_DEFAULT_FORMAT, self.AUDIO_CHANNELS, 1024)

        self.window = sdl2.SDL_CreateWindow(self.title,
                                            sdl2.SDL_WINDOWPOS_CENTERED,
                                            sdl2.SDL_WINDOWPOS_CENTERED,
                                            self.width * self.zoom,
                                            self.height * self.zoom,
                                            sdl2.SDL_WINDOW_HIDDEN
                                            )
        self.renderer = sdl2.SDL_CreateRenderer(self.window, -1, sdl2.SDL_RENDERER_ACCELERATED)
        self.renderer_obj = Renderer(self.renderer)

        if self.zoom != 1:
            sdl2.SDL_RenderSetScale(self.renderer, self.zoom, self.zoom)

    def _update(self, dt):

        self.dt += dt
        while self.dt >= self.UPDATE_DT:
            for update in self.update_handlers:
                update(self.UPDATE_DT)
            self.dt -= dt

    def _draw(self):

        for draw in self.draw_handlers:
            draw(self.renderer_obj)

    def quit(self):
        """Quits the game"""
        self._quit = True

    def loop(self):
        """The game loop"""
        sdl2.SDL_ShowWindow(self.window)

        current = sdl2.SDL_GetTicks()
        while not self._quit:
            event = sdl2.SDL_Event()
            while sdl2.SDL_PollEvent(ctypes.byref(event)) != 0:
                if event.type == sdl2.SDL_QUIT:
                    self._quit = True
                    break

            self.keys = sdl2.SDL_GetKeyboardState(None)

            new = sdl2.SDL_GetTicks()
            self._update(new - current)

            if new - current > self.DRAW_DT:
                sdl2.SDL_RenderClear(self.renderer)
                self._draw()
                sdl2.SDL_RenderPresent(self.renderer)

            current = new

        for resource in self.resources.copy().keys():
            self.free_resource(resource)

        sdl2.SDL_DestroyRenderer(self.renderer)
        sdl2.SDL_HideWindow(self.window)
        sdl2.SDL_DestroyWindow(self.window)

        sdlmixer.Mix_Quit()
        sdl2.SDL_Quit()

    def draw(self, fn):
        self.draw_handlers.append(fn)
        return fn

    def update(self, fn):
        self.update_handlers.append(fn)
        return fn

    def play(self, sample, loops=0):
        """Plays a sample loaded with load_resource"""
        return sdlmixer.Mix_PlayChannel(-1, sample, loops)

    def stop_playback(self, channel=-1):
        """Stops the audio playback"""
        return sdlmixer.Mix_HaltChannel(channel)

    def free_resource(self, filename):
        """Free resources"""

        try:
            free_fn = self.resources[filename]
        except KeyError:
            return

        free_fn()
        del self.resources[filename]

    def load_resource(self, filename):
        """Loads resources"""

        found_path = None
        for path in self.resource_path:
            full_path = os.path.realpath(os.path.join(path, filename))
            if os.path.isfile(full_path):
                found_path = full_path
                break

        if found_path is None:
            raise OSError("Resource not found")

        if filename[-4:] == ".bmp":
            image = sdl2.SDL_LoadBMP(found_path.encode())
            if not image:
                sys.exit("Error loading %r: %s" % (filename, sdl2.SDL_GetError()))

            resource = sdl2.SDL_CreateTextureFromSurface(self.renderer, image);
            free_fn = sdl2.SDL_DestroyTexture
            sdl2.SDL_FreeSurface(image)
        elif filename[-4:] in (".png", ".gif", ".jpg"):
            from sdl2 import sdlimage

            image = sdlimage.IMG_Load(found_path.encode())
            if not image:
                sys.exit("Error loading %r: %s" % (filename, sdlimage.IMG_GetError()))

            resource = sdl2.SDL_CreateTextureFromSurface(self.renderer, image);
            free_fn = sdl2.SDL_DestroyTexture
            sdl2.SDL_FreeSurface(image)
        elif filename[-4:] in (".wav", ".ogg"):
            audio = sdlmixer.Mix_LoadWAV(found_path.encode())
            if not audio:
                sys.exit("Error loading %r: %s" % (filename, sdlmixer.Mix_GetError()))

            resource = audio
            free_fn = sdlmixer.Mix_FreeChunk
        else:
            return open(filename, "rb")

        self.resources[filename] = lambda : free_fn(resource)
        return resource

    def load_bitmap_font(self, filename, width, height, font_map=None):
        """
        Loads a bitmap font.

        Parameters:

            filename: image containing the font (eg, font.png).
            width: wiidth of a font character.
            height: height of a font character.
            font_map: string with the order of the characters in the font.

        The default font map is:

        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!?()@:/'., "

        """
        if font_map is None:
            font_map = self.FONT_MAP

        font = BitmapFont(texture=self.load_resource(filename),
                          width=width,
                          height=height,
                          font_map=font_map,
                          )
        self.bitmap_fonts[filename] = font
        return font

class Renderer(object):
    """Wrapper for the renderer to be used by the draw functions"""
    def __init__(self, renderer):
        self.renderer = renderer

    def draw(self, texture, src_rect=None, dest_rect=None, tint=None):
        """
        Draw a texture.

        Parameters:

            texture: font name (load it first with load_bitmap_font).
            src_rect: tuple with the rect defining the section of the texture to draw.
            dest_rect: tuple with the rect defining the section of the destination.
            tint: colour the text texture, tuple with (r, g, b, alpha).
        """
        if isinstance(src_rect, tuple):
            src = sdl2.SDL_Rect(*src_rect)
        else:
            src = None

        if isinstance(dest_rect, tuple):
            dest = sdl2.SDL_Rect(*dest_rect)
        else:
            dest = None

        if isinstance(tint, tuple) and len(tint) == 4:
            sdl2.SDL_SetTextureColorMod(texture, *tint)
        else:
            tint = None

        sdl2.SDL_RenderCopy(self.renderer, texture, src, dest)

        if tint:
            sdl2.SDL_SetTextureColorMod(texture, 255, 255, 255, 255)

    def draw_text(self, font, x, y, text, align="left", tint=None):
        """
        Draw text using a texture.

        Parameters:

            font: font name (load it first with load_bitmap_font).
            x: horizontal position on the screen.
            y: vertical position on the screen.
            text: the text to render.
            align: "left", "right" or "center" (defaults to "left").
            tint: colour the text texture, tuple with (r, g, b, alpha).
        """
        width = len(text) * font.width

        if align == "center":
            x -= width // 2
            y -= font.height // 2
        elif align == "right":
            x -= width

        src = sdl2.SDL_Rect(0, 0, font.width, font.height)
        dest = sdl2.SDL_Rect(0, y, font.width, font.height)

        if isinstance(tint, tuple) and len(tint) == 4:
            sdl2.SDL_SetTextureColorMod(font.texture, *tint)
        else:
            tint = None

        for i, c in enumerate(text):
            index = font.font_map.find(c)
            src.x = index * font.width
            dest.x = x + i * font.width
            sdl2.SDL_RenderCopy(self.renderer, font.texture, src, dest)

        if tint:
            sdl2.SDL_SetTextureColorMod(font.texture, 255, 255, 255, 255)

BitmapFont = namedtuple("BitmapFont", ["texture", "width", "height", "font_map"])

